#44: Add a first draft of a guidelines PEP (#46)
capi-workgroup/api-evolution: guidelines.rst

PEP: 9999
Title: Guidelines for Python's C API
Author: Petr Viktorin <encukou@gmail.com>,
Discussions-To: https://discuss.python.org/c/c-api/30
Status: Draft
Type: Process
Requires: 731
Created: 20-Nov-2023
Post-History: XXX

.. pep-banner::

   This is a **incomplete draft**, published for initial discussion.
   It is likely to be amended substantially.

.. Authors to be added:

        Guido van Rossum <guido@python.org>,
        Victor Stinner <vstinner@python.org>,
        Steve Dower <steve.dower@python.org>,

>         Irit Katriel <irit@python.org>

Reviewer: Erlend?

Abstract
========

TODO: Write the abstract last :)

About this document

> ===================

Reviewer 2: It seems to me the _About this document_ section could function as an _Abstract_ :)

Author: Still, I think it should be moved there last :)

> These guidelines represent the consensus of the C API working group,

Reviewer 2: ```suggestion
These guidelines represent the consensus of the C API Working Group,
```

established in :pep:`731` to oversee and coordinate

> the development and maintenance of the Python C API.

Reviewer 2: CPython?

```suggestion
the development and maintenance of the CPython C API.
```

It is a living document. It can be changed with the approval of either

> all members of the C API working group, or the Steering Council.

Reviewer 2: ```suggestion
all members of the C API Working Group, or the Steering Council.
```

High-level goals

> ================

Reviewer 2: I guess for a PEP, this would be a subheading under the _Specification_ section. Perhaps we should establish a _Speficiation_ heading and put the relevant subheadings under it.

Author: This is not a standards track PEP. Is *Specification* appropriate for guidelines?

Reviewer 2: Well, _Specification_ was used for the Standards Track PEP-689, so I guess it can be used for guidelines ;)

Reviewer 2: We also used a _Specification_ section for the Standards Track PEP-687.

Reviewer 2: Right; I just realised this is a _process_ PEP 😄

Author: FWIW, it also already says:
> It is a living document. It can be changed with the approval of either
> all members of the C API Working Group, or the Steering Council.

Our goal is that the C API is:

* **Useful**: It should provide access to all relevant functionality.
* **Ergonomic**: It should be easy to use.

> * **Fair**: It should serve the purposes of a variety of stakeholders.

Author: There's a lot to add :)

* **Stable**: It should avoid requiring users to update their code more than necessary.
* **Evolvable**: It should give Python implementers enough freedom to change implementation details.
* **Maintainable**: It should be easy to maintain and extend.

.. note:: 

    For background and discussions, see:

    - https://github.com/capi-workgroup/api-evolution/issues/26

> Meta-guidelines

Author: I thought the [*meta* prefix](https://en.wikipedia.org/wiki/Meta_(prefix)#Epistemology) makes for a good way to spell “Guidelines for using these guidelines”.

I don't have a better proposal now, so let's keep this to a future PR.

In several cases, you should seek approval from the C API working group
before changing public C API.
Most notably:

* when the guidelines tell you to do so,
* when adding an exception to the rules,
* when adding to the limited API,
* when adding something that is likely to establish precedent,
* when adding many APIs at once, or
* when the guidelines are unclear or don't make sense.

Approvals allow the working group to coordinate designs from multiple people,
and to collect feedback needed to refine these guidelines.

Request an approval by opening `an issue in the decisions repository`_.

.. _an issue in the decisions repository: https://github.com/capi-workgroup/decisions/issues

Exceptions to the guidelines
----------------------------

Where possible, exceptions to these guidelines should be added as
*alternatives* to API that follows the guidelines.
Such exceptions should be clearly identifiable from their names.

If you want to add an exception that is not mentioned in the guidelines,
please request approval from the C API workgroup.

.. note:: 

    For background and discussions, see:

    - https://github.com/capi-workgroup/api-evolution/issues/4

Applicability

> =============

Author: I'd rather keep the individual sections (like “types”, “return values”, “managing ownership”) at the top level. Mixing in the few other sections (like “Abstract”, “About this document”, “Copyright”) doesn't really hurt, and with some (“Meta-guidelines”) it's unclear if they should be in “C API guidelines” or outside.

The PEP is named “Guidelines for Python's C API” already :)

These guidelines only apply to API that is:

* *Newly added*. Existing API does not necessarily follow these guidelines;
  this PEP is not the place for plans to replace, deprecate or remove it.

> * *Pubic*: meant for users. That is, not in the *private* or *internal*

Reviewer 3: ```suggestion
* *Public*: meant for users. That is, not in the *private* or *internal*
```

  tiers as defined in the next section.

Note that we have a style guide, :pep:`7`, which applies to *all* C code
in CPython, including the API.

API tiers
---------

Python has several tiers of C API.
The more stable the tier is, the stricter the rules for it are:

> *  For **internal** API (only accessible with ``Py_BUILD_CORE``, or not declared

Author: I don't see the benefit of the more complex syntax.

> *  For **internal** API (only accessible with ``Py_BUILD_CORE``, or not declared

Reviewer 2: It renders nicer.

> *  For **internal** API (only accessible with ``Py_BUILD_CORE``, or not declared

Author: AFAICS, it renders exactly the same.
I'd not object to going all the way and switching to Intersphinx, but using the role and exclamation mark everywhere is just extra typing. We don't get links, nor typo warnings (for that we'd also keep to the PEP in sync with the `cpython` repo), nor even syntax checks (AFAIK, some of your suggestions here are wrong, you can't put the star in ```:c:type:`PyObject*` ``` -- but I'm pretty sure I'd make the same kind of mistakes without a checker).

> *  For **internal** API (only accessible with ``Py_BUILD_CORE``, or not declared

Reviewer 2: Fair enough.

   in a public header at all), you are free to ignore these guidelines entirely.

*  **Private** API (not intended for users) should be *internal* if possible.
   Otherwise it must use the ``_Py`` prefix (see :ref:`naming`).
   The rest of these guidelines don't apply to it.

*  In **unstable** API (prefixed with ``PyUnstable_``), it is OK to ignore
   these guidelines if there is a reason to do so -- usually for performance.

>    Please document the reason in a source comment.

Author: Or maybe the Devguide rather than the historical document?

>    Please document the reason in a source comment.

Reviewer 2: Yes, makes sense. We can link to https://devguide.python.org/developer-workflow/c-api/#unstable-c-api.

>    Please document the reason in a source comment.

Reviewer 2: Let's link to the historical document (PEP-689) for now.

*  In the **general** public API, these guidelines apply in full force.

>    The C API working group can grant exceptions.

Author: It would be nice to find a good term here. Unstable API is also public.

>    The C API working group can grant exceptions.

Reviewer 2: The "Public API" is what we're currently using both in the docs and in the devguide.

>    The C API working group can grant exceptions.

Reviewer 4: The Limited API, CPython API, and ...un-limited(?) API are all also public, so there's probably a definition required somewhere to define the scope. I don't see a problem bundling them all together for this statement (though it may read more friendly if we say something like "assume that all guidelines apply, and seek advice from the C API WG if an exception may be necessary or beneficial")

>    The C API working group can grant exceptions.

Author: Devguide uses *The “general” public C API* in its [list of the four tiers](https://devguide.python.org/developer-workflow/c-api/index.html).
[Docs](https://docs.python.org/3/c-api/stable.html) explain this “default” tier, and then carve out the unstable & limited tiers with different stability expectations.

(I wrote both of those; this is the best wording I found.)

*  For the **limited** API, please get explicit approval from the
   C API working group for any changes.

.. note:: 

    For background and discussions, see:

    - https://github.com/capi-workgroup/api-evolution/issues/42 (for limited API)

Types
=====

Objects
-------

> Use ``PyObject*`` for all Python objects.
> Avoid using concrete types (e.g. ``PyDictObject*``).

Reviewer 2: ```suggestion
Use :c:type:`!PyObject*` for all Python objects.
Avoid using concrete types, e.g. :c:type:`PyDictObject*`.
```

Public API should type-check all objects passed to it.
When it gets an object of an unexpected type, public API should fail with
``TypeError`` rather than crash.

> As an exception, with approval from the C API WG you can use concrete types,

Reviewer 2: ```suggestion
As an exception, with approval from the C API Working Group you can use concrete types,
```

> such as ``PyTypeObject*``, ``PyCodeObject*`` & ``PyFrameObject*``,
> for consistency with existing API.
> These objects should be type-checked as if they were ``PyObject*``.

Reviewer 2: ```suggestion
such as :c:type:`!PyTypeObject*`, :c:type:`!PyCodeObject*` and :c:type:`!PyFrameObject*`,
for consistency with existing API.
These objects should be type-checked as if they were :c:type:`!PyObject*`.
```
